File: qucanft/zodiac.py
"""
Zodiac calculations module for converting celestial coordinates to zodiac signs.

This module provides functionality to calculate zodiac signs, degrees, and 
symbolic representations from astronomical coordinates.
"""

from typing import Dict, List, Optional, Tuple
import numpy as np
import pandas as pd
from datetime import datetime
from astropy.coordinates import SkyCoord
from astropy import units as u


class ZodiacCalculator:
    """
    A class for calculating zodiac signs and positions from astronomical coordinates.
    
    This class converts Right Ascension and Declination coordinates to zodiac
    positions and provides symbolic representations.
    """
    
    # Zodiac signs with their celestial longitude ranges (in degrees)
    ZODIAC_SIGNS = [
        {'name': 'Aries', 'symbol': '♈', 'element': 'Fire', 'quality': 'Cardinal', 'ruler': 'Mars'},
        {'name': 'Taurus', 'symbol': '♉', 'element': 'Earth', 'quality': 'Fixed', 'ruler': 'Venus'},
        {'name': 'Gemini', 'symbol': '♊', 'element': 'Air', 'quality': 'Mutable', 'ruler': 'Mercury'},
        {'name': 'Cancer', 'symbol': '♋', 'element': 'Water', 'quality': 'Cardinal', 'ruler': 'Moon'},
        {'name': 'Leo', 'symbol': '♌', 'element': 'Fire', 'quality': 'Fixed', 'ruler': 'Sun'},
        {'name': 'Virgo', 'symbol': '♍', 'element': 'Earth', 'quality': 'Mutable', 'ruler': 'Mercury'},
        {'name': 'Libra', 'symbol': '♎', 'element': 'Air', 'quality': 'Cardinal', 'ruler': 'Venus'},
        {'name': 'Scorpio', 'symbol': '♏', 'element': 'Water', 'quality': 'Fixed', 'ruler': 'Mars'},
        {'name': 'Sagittarius', 'symbol': '♐', 'element': 'Fire', 'quality': 'Mutable', 'ruler': 'Jupiter'},
        {'name': 'Capricorn', 'symbol': '♑', 'element': 'Earth', 'quality': 'Cardinal', 'ruler': 'Saturn'},
        {'name': 'Aquarius', 'symbol': '♒', 'element': 'Air', 'quality': 'Fixed', 'ruler': 'Saturn'},
        {'name': 'Pisces', 'symbol': '♓', 'element': 'Water', 'quality': 'Mutable', 'ruler': 'Jupiter'}
    ]
    
    def __init__(self):
        """Initialize the ZodiacCalculator."""
        pass
    
    def ra_dec_to_ecliptic(self, ra: float, dec: float, epoch: str = 'J2000') -> Tuple[float, float]:
        """
        Convert Right Ascension and Declination to ecliptic coordinates.
        
        Args:
            ra: Right Ascension in degrees
            dec: Declination in degrees
            epoch: Epoch for the coordinate system (default: J2000)
            
        Returns:
            Tuple of (ecliptic_longitude, ecliptic_latitude) in degrees
        """
        # Create SkyCoord object
        coord = SkyCoord(ra=ra*u.degree, dec=dec*u.degree, frame='icrs')
        
        # Convert to ecliptic coordinates
        ecliptic_coord = coord.geocentrictrueecliptic
        
        return ecliptic_coord.lon.degree, ecliptic_coord.lat.degree
    
    def ecliptic_to_zodiac(self, ecliptic_longitude: float) -> Dict[str, any]:
        """
        Convert ecliptic longitude to zodiac sign and degree.
        
        Args:
            ecliptic_longitude: Ecliptic longitude in degrees
            
        Returns:
            Dictionary with zodiac information including sign, degree, symbol, etc.
        """
        # Normalize longitude to 0-360 range
        longitude = ecliptic_longitude % 360
        
        # Calculate zodiac sign (each sign is 30 degrees)
        sign_index = int(longitude // 30)
        degree_in_sign = longitude % 30
        
        # Get zodiac sign information
        zodiac_info = self.ZODIAC_SIGNS[sign_index].copy()
        zodiac_info.update({
            'degree': degree_in_sign,
            'absolute_degree': longitude,
            'degree_minutes': int((degree_in_sign % 1) * 60),
            'degree_seconds': int(((degree_in_sign % 1) * 60 % 1) * 60),
            'position_string': f"{int(degree_in_sign)}°{int((degree_in_sign % 1) * 60):02d}' {zodiac_info['symbol']}"
        })
        
        return zodiac_info
    
    def calculate_zodiac_positions(self, planetary_data: pd.DataFrame) -> pd.DataFrame:
        """
        Calculate zodiac positions for planetary data.
        
        Args:
            planetary_data: DataFrame with planetary positions (must have 'RA' and 'Dec' columns)
            
        Returns:
            DataFrame with added zodiac position columns
        """
        if 'RA' not in planetary_data.columns or 'Dec' not in planetary_data.columns:
            raise ValueError("DataFrame must contain 'RA' and 'Dec' columns")
        
        result_df = planetary_data.copy()
        
        # Calculate zodiac positions for each planet
        zodiac_data = []
        
        for _, row in planetary_data.iterrows():
            # Convert RA/Dec to ecliptic coordinates
            ecl_lon, ecl_lat = self.ra_dec_to_ecliptic(row['RA'], row['Dec'])
            
            # Get zodiac information
            zodiac_info = self.ecliptic_to_zodiac(ecl_lon)
            
            zodiac_data.append({
                'Ecliptic_Longitude': ecl_lon,
                'Ecliptic_Latitude': ecl_lat,
                'Zodiac_Sign': zodiac_info['name'],
                'Zodiac_Symbol': zodiac_info['symbol'],
                'Zodiac_Degree': zodiac_info['degree'],
                'Zodiac_Element': zodiac_info['element'],
                'Zodiac_Quality': zodiac_info['quality'],
                'Zodiac_Ruler': zodiac_info['ruler'],
                'Position_String': zodiac_info['position_string']
            })
        
        # Add zodiac columns to the result DataFrame
        zodiac_df = pd.DataFrame(zodiac_data)
        result_df = pd.concat([result_df, zodiac_df], axis=1)
        
        return result_df
    
    def get_zodiac_sign_info(self, sign_name: str) -> Optional[Dict[str, any]]:
        """
        Get detailed information about a zodiac sign.
        
        Args:
            sign_name: Name of the zodiac sign
            
        Returns:
            Dictionary with zodiac sign information, or None if not found
        """
        for sign in self.ZODIAC_SIGNS:
            if sign['name'].lower() == sign_name.lower():
                return sign.copy()
        return None
    
    def get_zodiac_compatibility(self, sign1: str, sign2: str) -> Dict[str, any]:
        """
        Calculate basic zodiac compatibility based on elements and qualities.
        
        Args:
            sign1: First zodiac sign name
            sign2: Second zodiac sign name
            
        Returns:
            Dictionary with compatibility information
        """
        info1 = self.get_zodiac_sign_info(sign1)
        info2 = self.get_zodiac_sign_info(sign2)
        
        if not info1 or not info2:
            return {'compatibility': 'Unknown', 'reason': 'Invalid zodiac sign'}
        
        # Element compatibility
        element_compatibility = {
            ('Fire', 'Fire'): 'High',
            ('Fire', 'Air'): 'High',
            ('Fire', 'Water'): 'Low',
            ('Fire', 'Earth'): 'Medium',
            ('Earth', 'Earth'): 'High',
            ('Earth', 'Water'): 'High',
            ('Earth', 'Air'): 'Low',
            ('Air', 'Air'): 'High',
            ('Air', 'Water'): 'Low',
            ('Water', 'Water'): 'High'
        }
        
        element_pair = (info1['element'], info2['element'])
        if element_pair not in element_compatibility:
            element_pair = (info2['element'], info1['element'])
        
        element_comp = element_compatibility.get(element_pair, 'Medium')
        
        # Quality compatibility
        quality_compatibility = {
            ('Cardinal', 'Cardinal'): 'Medium',
            ('Cardinal', 'Fixed'): 'Low',
            ('Cardinal', 'Mutable'): 'High',
            ('Fixed', 'Fixed'): 'Low',
            ('Fixed', 'Mutable'): 'Medium',
            ('Mutable', 'Mutable'): 'High'
        }
        
        quality_pair = (info1['quality'], info2['quality'])
        if quality_pair not in quality_compatibility:
            quality_pair = (info2['quality'], info1['quality'])
        
        quality_comp = quality_compatibility.get(quality_pair, 'Medium')
        
        # Overall compatibility
        comp_scores = {'High': 3, 'Medium': 2, 'Low': 1}
        avg_score = (comp_scores[element_comp] + comp_scores[quality_comp]) / 2
        
        if avg_score >= 2.5:
            overall = 'High'
        elif avg_score >= 1.5:
            overall = 'Medium'
        else:
            overall = 'Low'
        
        return {
            'sign1': sign1,
            'sign2': sign2,
            'element_compatibility': element_comp,
            'quality_compatibility': quality_comp,
            'overall_compatibility': overall,
            'description': f"{sign1} ({info1['element']}, {info1['quality']}) and {sign2} ({info2['element']}, {info2['quality']})"
        }
    
    def calculate_midpoint(self, pos1: float, pos2: float) -> float:
        """
        Calculate the midpoint between two zodiac positions.
        
        Args:
            pos1: First position in degrees (0-360)
            pos2: Second position in degrees (0-360)
            
        Returns:
            Midpoint in degrees (0-360)
        """
        # Normalize positions
        pos1 = pos1 % 360
        pos2 = pos2 % 360
        
        # Calculate shortest arc between positions
        diff = abs(pos2 - pos1)
        if diff > 180:
            diff = 360 - diff
            # Use the longer arc
            midpoint = (pos1 + pos2 + 360) / 2
        else:
            midpoint = (pos1 + pos2) / 2
        
        return midpoint % 360
    
    def degrees_to_dms(self, degrees: float) -> Tuple[int, int, int]:
        """
        Convert decimal degrees to degrees, minutes, seconds.
        
        Args:
            degrees: Decimal degrees
            
        Returns:
            Tuple of (degrees, minutes, seconds)
        """
        deg = int(degrees)
        min_float = (degrees - deg) * 60
        minutes = int(min_float)
        seconds = int((min_float - minutes) * 60)
        
        return deg, minutes, seconds
    
    def format_zodiac_position(self, longitude: float, include_symbol: bool = True) -> str:
        """
        Format a zodiac position as a readable string.
        
        Args:
            longitude: Ecliptic longitude in degrees
            include_symbol: Whether to include the zodiac symbol
            
        Returns:
            Formatted position string
        """
        zodiac_info = self.ecliptic_to_zodiac(longitude)
        
        if include_symbol:
            return zodiac_info['position_string']
        else:
            deg, min_val, sec = self.degrees_to_dms(zodiac_info['degree'])
            return f"{deg}°{min_val:02d}'{sec:02d}\" {zodiac_info['name']}"

File: pkg/zodiac/zodiac.go
// Package zodiac provides zodiac sign calculations and interpretations.
package zodiac

import (
	"fmt"
	"math"
	
	"github.com/Qucanft/Qucanft/pkg/coordinates"
)

// ZodiacSign represents a zodiac sign
type ZodiacSign struct {
	Name     string
	Symbol   string
	Element  string
	Quality  string
	Ruler    string
	StartDeg float64 // Starting degree in the ecliptic
	EndDeg   float64 // Ending degree in the ecliptic
}

// ZodiacPosition represents a position within a zodiac sign
type ZodiacPosition struct {
	Sign        ZodiacSign
	DegreeInSign float64 // 0-30 degrees within the sign
	AbsoluteDeg  float64 // 0-360 degrees absolute position
}

// ZodiacCalculator handles zodiac-related calculations
type ZodiacCalculator struct {
	signs []ZodiacSign
}

// NewZodiacCalculator creates a new zodiac calculator
func NewZodiacCalculator() *ZodiacCalculator {
	return &ZodiacCalculator{
		signs: getZodiacSigns(),
	}
}

// GetZodiacSigns returns all zodiac signs
func (zc *ZodiacCalculator) GetZodiacSigns() []ZodiacSign {
	return zc.signs
}

// GetSignByName returns a zodiac sign by name
func (zc *ZodiacCalculator) GetSignByName(name string) (ZodiacSign, bool) {
	for _, sign := range zc.signs {
		if sign.Name == name {
			return sign, true
		}
	}
	return ZodiacSign{}, false
}

// EclipticToZodiac converts ecliptic longitude to zodiac position
func (zc *ZodiacCalculator) EclipticToZodiac(longitude float64) ZodiacPosition {
	// Normalize longitude to 0-360 range
	longitude = coordinates.NormalizeAngle(longitude)
	
	// Find the zodiac sign
	signIndex := int(longitude / 30.0)
	if signIndex >= len(zc.signs) {
		signIndex = len(zc.signs) - 1
	}
	
	sign := zc.signs[signIndex]
	degreeInSign := longitude - sign.StartDeg
	
	return ZodiacPosition{
		Sign:        sign,
		DegreeInSign: degreeInSign,
		AbsoluteDeg:  longitude,
	}
}

// ZodiacToEcliptic converts zodiac position back to ecliptic longitude
func (zc *ZodiacCalculator) ZodiacToEcliptic(position ZodiacPosition) float64 {
	return position.Sign.StartDeg + position.DegreeInSign
}

// GetSignCompatibility returns compatibility score between two signs (0-100)
func (zc *ZodiacCalculator) GetSignCompatibility(sign1, sign2 ZodiacSign) float64 {
	// Simple compatibility based on elements and qualities
	elementScore := 0.0
	qualityScore := 0.0
	
	// Element compatibility
	if sign1.Element == sign2.Element {
		elementScore = 40.0 // Same element
	} else if isCompatibleElement(sign1.Element, sign2.Element) {
		elementScore = 30.0 // Compatible elements
	} else {
		elementScore = 10.0 // Less compatible elements
	}
	
	// Quality compatibility
	if sign1.Quality == sign2.Quality {
		qualityScore = 20.0 // Same quality
	} else if isCompatibleQuality(sign1.Quality, sign2.Quality) {
		qualityScore = 25.0 // Compatible qualities
	} else {
		qualityScore = 15.0 // Less compatible qualities
	}
	
	// Angular relationship bonus
	angle := math.Abs(sign1.StartDeg - sign2.StartDeg)
	if angle > 180 {
		angle = 360 - angle
	}
	
	angleScore := 0.0
	switch {
	case angle == 0:
		angleScore = 35.0 // Same sign
	case angle == 60 || angle == 120:
		angleScore = 35.0 // Trine or sextile
	case angle == 90 || angle == 180:
		angleScore = 15.0 // Square or opposition
	default:
		angleScore = 25.0 // Other angles
	}
	
	return elementScore + qualityScore + angleScore
}

// CalculateAspectAngle calculates the aspect angle between two zodiac positions
func (zc *ZodiacCalculator) CalculateAspectAngle(pos1, pos2 ZodiacPosition) float64 {
	angle := math.Abs(pos1.AbsoluteDeg - pos2.AbsoluteDeg)
	if angle > 180 {
		angle = 360 - angle
	}
	return angle
}

// FormatZodiacPosition formats a zodiac position as a string
func (zc *ZodiacCalculator) FormatZodiacPosition(position ZodiacPosition) string {
	degrees := int(position.DegreeInSign)
	minutes := int((position.DegreeInSign - float64(degrees)) * 60)
	seconds := int(((position.DegreeInSign - float64(degrees)) * 60 - float64(minutes)) * 60)
	
	return fmt.Sprintf("%d°%d'%d\" %s", degrees, minutes, seconds, position.Sign.Name)
}

// IsRetrograde determines if a planet appears to be in retrograde motion
// This is a simplified calculation based on orbital mechanics
func (zc *ZodiacCalculator) IsRetrograde(planetName string, longitude1, longitude2 float64, timeDiff float64) bool {
	// Calculate apparent motion
	motion := (longitude2 - longitude1) / timeDiff
	
	// Normalize motion
	if motion > 180 {
		motion -= 360
	} else if motion < -180 {
		motion += 360
	}
	
	// Different planets have different retrograde thresholds
	threshold := getRetrogradeThreshold(planetName)
	
	return motion < threshold
}

// getRetrogradeThreshold returns the retrograde motion threshold for a planet
func getRetrogradeThreshold(planetName string) float64 {
	thresholds := map[string]float64{
		"Mercury": -0.5,
		"Venus":   -0.3,
		"Mars":    -0.2,
		"Jupiter": -0.1,
		"Saturn":  -0.05,
		"Uranus":  -0.02,
		"Neptune": -0.01,
		"Pluto":   -0.008,
	}
	
	if threshold, exists := thresholds[planetName]; exists {
		return threshold
	}
	
	return -0.1 // Default threshold
}

// isCompatibleElement checks if two elements are compatible
func isCompatibleElement(element1, element2 string) bool {
	compatible := map[string][]string{
		"Fire":  {"Air"},
		"Earth": {"Water"},
		"Air":   {"Fire"},
		"Water": {"Earth"},
	}
	
	if compatibleElements, exists := compatible[element1]; exists {
		for _, elem := range compatibleElements {
			if elem == element2 {
				return true
			}
		}
	}
	
	return false
}

// isCompatibleQuality checks if two qualities are compatible
func isCompatibleQuality(quality1, quality2 string) bool {
	// Cardinal and mutable are generally compatible
	// Fixed signs can be compatible with both but with more challenge
	compatible := map[string][]string{
		"Cardinal": {"Mutable"},
		"Fixed":    {"Cardinal", "Mutable"},
		"Mutable":  {"Cardinal"},
	}
	
	if compatibleQualities, exists := compatible[quality1]; exists {
		for _, qual := range compatibleQualities {
			if qual == quality2 {
				return true
			}
		}
	}
	
	return false
}

// getZodiacSigns returns the 12 zodiac signs with their properties
func getZodiacSigns() []ZodiacSign {
	return []ZodiacSign{
		{
			Name:     "Aries",
			Symbol:   "♈",
			Element:  "Fire",
			Quality:  "Cardinal",
			Ruler:    "Mars",
			StartDeg: 0,
			EndDeg:   30,
		},
		{
			Name:     "Taurus",
			Symbol:   "♉",
			Element:  "Earth",
			Quality:  "Fixed",
			Ruler:    "Venus",
			StartDeg: 30,
			EndDeg:   60,
		},
		{
			Name:     "Gemini",
			Symbol:   "♊",
			Element:  "Air",
			Quality:  "Mutable",
			Ruler:    "Mercury",
			StartDeg: 60,
			EndDeg:   90,
		},
		{
			Name:     "Cancer",
			Symbol:   "♋",
			Element:  "Water",
			Quality:  "Cardinal",
			Ruler:    "Moon",
			StartDeg: 90,
			EndDeg:   120,
		},
		{
			Name:     "Leo",
			Symbol:   "♌",
			Element:  "Fire",
			Quality:  "Fixed",
			Ruler:    "Sun",
			StartDeg: 120,
			EndDeg:   150,
		},
		{
			Name:     "Virgo",
			Symbol:   "♍",
			Element:  "Earth",
			Quality:  "Mutable",
			Ruler:    "Mercury",
			StartDeg: 150,
			EndDeg:   180,
		},
		{
			Name:     "Libra",
			Symbol:   "♎",
			Element:  "Air",
			Quality:  "Cardinal",
			Ruler:    "Venus",
			StartDeg: 180,
			EndDeg:   210,
		},
		{
			Name:     "Scorpio",
			Symbol:   "♏",
			Element:  "Water",
			Quality:  "Fixed",
			Ruler:    "Mars",
			StartDeg: 210,
			EndDeg:   240,
		},
		{
			Name:     "Sagittarius",
			Symbol:   "♐",
			Element:  "Fire",
			Quality:  "Mutable",
			Ruler:    "Jupiter",
			StartDeg: 240,
			EndDeg:   270,
		},
		{
			Name:     "Capricorn",
			Symbol:   "♑",
			Element:  "Earth",
			Quality:  "Cardinal",
			Ruler:    "Saturn",
			StartDeg: 270,
			EndDeg:   300,
		},
		{
			Name:     "Aquarius",
			Symbol:   "♒",
			Element:  "Air",
			Quality:  "Fixed",
			Ruler:    "Uranus",
			StartDeg: 300,
			EndDeg:   330,
		},
		{
			Name:     "Pisces",
			Symbol:   "♓",
			Element:  "Water",
			Quality:  "Mutable",
			Ruler:    "Neptune",
			StartDeg: 330,
			EndDeg:   360,
		},
	}
}

// String methods
func (zs ZodiacSign) String() string {
	return fmt.Sprintf("%s (%s) - %s %s, ruled by %s", zs.Name, zs.Symbol, zs.Element, zs.Quality, zs.Ruler)
}

func (zp ZodiacPosition) String() string {
	return fmt.Sprintf("%.2f° %s (%.2f° absolute)", zp.DegreeInSign, zp.Sign.Name, zp.AbsoluteDeg)
}